src/cf.data.pre

```diff
@@ -1275,9 +1275,24 @@ ENDIF
 	  # Referer is highly unreliable, so use with care
 
 	acl aclname ident [-i] username ...
+	  # string match on the ident user identifier [slow]
+	  # use REQUIRED to accept any non-null user identifier.
 	acl aclname ident_regex [-i] pattern ...
-	  # string match on ident output [slow]
-	  # use REQUIRED to accept any non-null ident.
+	  # regex match against the ident user identifier [slow]
+	  #
+	  # If necessary, evaluation of this ACL triggers a transaction with an
+	  # Authentication Server (a.k.a. Ident server; see RFC 1413) to identify
+	  # the user behind client-to-Squid TCP connection. The results of that
+	  # transaction (successful or otherwise) are cached for the lifetime the
+	  # client-to-Squid TCP connection at most one attempt to identify connection.
+	  # user is made.
+	  #
+	  # A configured dash ('-') value (or any equivalent regex) matches:
+	  #    - A connection for which Squid has failed to obtain user identity,
+	  #      including cases where Authentication Server returned an invalidd
+	  #      (e.g., empty) user-id.
+	  #    - A connection for which Authentication Server returned a dash character
+	  #      as a user-id.
```

Comment:
> I posted an adjusted version at [9841f32](https://github.com/measurement-factory/squid/commit/9841f32554862e9d8136a300d7ecd8992fb2bc31) without the 1st and 2nd of your listed '-' cases. In both cases checklist->ident() is undefined and it does not attempt to match().

> 1. A transaction that lacks access to a client-to-Squid connection. This case results in level 1 message"ERROR: Cannot start ident lookup" and ACCESS_DUNNO.
> 
> 2. A connection for which Squid has not obtained user identity yet. A fast ACL makes hecklist->goAsync() fail, resulting in level 3 message "cannot start ident lookup" and ACCESS_DUNNO.
 
> IMO the both cases indicate some general failure and we should not adjust the code to make them match '-'.

I am OK with keeping these two cases outside '-' reach, at least for now. However, I think we should add them to this documentation, as a separate list of problematic cases where this ACL does _not_ match. The admin should be aware of all these cases when writing security-sensitive rules.

When an ACL `foo` returns ACCESS_DUNNO, does `!foo` return ACCESS_DUNNO as well? What about an `any-of` and `all-of` ACLs? Will they return ACCESS_DUNNO if at least one of their nodes returns ACCESS_DUNNO?

Reply:
> !foo return ACCESS_DUNNO as well? 

Yes.

> What about an any-of and all-of ACLs?

any-of: ACCESS_DUNNO if ACCESS_DUNNO is returned by the first ACL. It will be ALLOWED, e.g., if the first ACL returns ALLOWED and the second ACL returns ACCESS_DUNNO.
all-of: ACCESS_DUNNO if one of the ACLs returns ACCESS_DUNNO.

Reply:
> any-of: ACCESS_DUNNO if ACCESS_DUNNO is returned by the first ACL. 

OK.

> It will be ALLOWED, e.g., if the first ACL returns ALLOWED and the second ACL returns ACCESS_DUNNO.

If the first any-of ACL node returns ALLOWED, do we really evaluate the second node? I would expect the evaluation to stop once the result is known.

If we are only looking at allow/deny/dunno outcomes, then I would expect any-of to result in 

* ALLOWED if at least one ACL node returned ALLOWED,
* DENIED if all ACL nodes returned DENIED,
* DUNNO otherwise (i.e. at least one node returned DUNNO and no nodes returned ALLOWED)

Do my expectations match reality?

Reply:
> we should add them to this documentation, as a separate list of problematic cases

Done (42e71c0)

Reply:
> If the first any-of ACL node returns ALLOWED, do we really evaluate the second node? 

No. Also, my test showed that if the first any-of ACL returns DUNNO and the second - ALLOWED, we also do not evaluate the second rule and return DUNNO.

> ALLOWED if at least one ACL node returned ALLOWED,

Yes for ALLOWED/DENIED and No for ALLOWED/DUNNO, e.g., ALLOWED,DUNNO results in ALLOWED and DUNNO,ALLOWED results in DUNNO (the second ACL in both cases is not considered).

> DUNNO (i.e. at least one node returned DUNNO and no nodes returned ALLOWED)

Yes, both DENIED,DUNNO and DUNNO,DENIED result in DUNNO.

Reply:
I sense a terminology conflict in our last few exchanges:

* When I say "ACL X returned Y", I mean "Evaluation of ACL X resulted in Y". In other words, I am talking about what is actually happening during Checklist application to some Squid directive.

* When you read my "ACL X returned Y", you are probably thinking "An evaluation of ACL X would result in Y". In other words, you are thinking about how that ACL X is written/configured.

For example, my "ALLOWED if at least one ACL node returned ALLOWED" means "ALLOWED if at least one _evaluated_ ACL node returned ALLOWED" rather than "ALLOWED if at least one ACL node would have returned ALLOWED if it were evaluated".

Here is the same expectation adjusted to make the above clarification explicit:

----

I expect ACLs evaluation to stop once the result is known. If we are only looking at allow/deny/dunno outcomes, then I would expect any-of evaluation to result in:

* ALLOWED if an evaluated ACL node returned ALLOWED,
* DENIED if all evaluated ACL nodes returned DENIED,
* DUNNO otherwise (i.e. an evaluated ACL returned DUNNO and no previously evaluated ACLs returned ALLOWED)

Do you agree that my bullets match the current official code?

Reply:
I did not re-run the tests, but IIRC once a node results in ALLOWED or DUNNO, the entire any-of ACL returns in that result:

> ALLOWED,DUNNO results in ALLOWED and DUNNO,ALLOWED results in DUNNO (the second ACL in both cases is not considered).

This goes in line with your (1) and (3) bullets.

> i.e. an evaluated ACL returned DUNNO and no previously evaluated ACLs returned ALLOWED

or "an evaluated ACL returned DUNNO and all previously evaluated ACLs returned DENIED".
It looks like (2) bullet is the product of (1) and (3).

Reply:
>  It looks like (2) bullet is the product of (1) and (3).

Yes, it is. Moreover, here, _any_ one bullet is the product of the other two because these three mutually exclusive bullets cover all possible in-scope cases. We can move that "X otherwise" shortcut phrase (with the right X) from the third bullet to any other bullet.